compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -4244,9 +4286,85 @@ object Types {
+              case tpnme.Times    if nArgs == 2 => constantFold2(floatValue, _ * _)
+              case tpnme.Div if nArgs == 2 => constantFold2(floatValue, _ / _)
+              case tpnme.Mod if nArgs == 2 => constantFold2(floatValue, _ % _)
+              case tpnme.Lt  if nArgs == 2 => constantFold2(floatValue, _ < _)
+              case tpnme.Gt  if nArgs == 2 => constantFold2(floatValue, _ > _)
+              case tpnme.Ge  if nArgs == 2 => constantFold2(floatValue, _ >= _)
+              case tpnme.Le  if nArgs == 2 => constantFold2(floatValue, _ <= _)
+              case tpnme.Min if nArgs == 2 => constantFold2(floatValue, _ min _)
+              case tpnme.Max if nArgs == 2 => constantFold2(floatValue, _ max _)
+              case tpnme.ToInt if nArgs == 1 => constantFold1(floatValue, _.toInt)
+              case tpnme.ToLong if nArgs == 1 => constantFold1(floatValue, _.toLong)
+              case tpnme.ToDouble if nArgs == 1 => constantFold1(floatValue, _.toDouble)
+              case _ => None
+            } else if (owner == defn.CompiletimeOpsDoubleModuleClass) name match {
+              case tpnme.Abs      if nArgs == 1 => constantFold1(doubleValue, _.abs)
+              case tpnme.Negate   if nArgs == 1 => constantFold1(doubleValue, x => -x)
+              case tpnme.Plus     if nArgs == 2 => constantFold2(doubleValue, _ + _)
+              case tpnme.Minus    if nArgs == 2 => constantFold2(doubleValue, _ - _)
+              case tpnme.Times    if nArgs == 2 => constantFold2(doubleValue, _ * _)
+              case tpnme.Div if nArgs == 2 => constantFold2(doubleValue, _ / _)
+              case tpnme.Mod if nArgs == 2 => constantFold2(doubleValue, _ % _)
+              case tpnme.Lt  if nArgs == 2 => constantFold2(doubleValue, _ < _)
+              case tpnme.Gt  if nArgs == 2 => constantFold2(doubleValue, _ > _)
+              case tpnme.Ge  if nArgs == 2 => constantFold2(doubleValue, _ >= _)
+              case tpnme.Le  if nArgs == 2 => constantFold2(doubleValue, _ <= _)
+              case tpnme.Min if nArgs == 2 => constantFold2(doubleValue, _ min _)
+              case tpnme.Max if nArgs == 2 => constantFold2(doubleValue, _ max _)
+              case tpnme.ToInt if nArgs == 1 => constantFold1(doubleValue, _.toInt)
+              case tpnme.ToLong if nArgs == 1 => constantFold1(doubleValue, _.toLong)
+              case tpnme.ToFloat if nArgs == 1 => constantFold1(doubleValue, _.toFloat)
               case _ => None
             } else if (owner == defn.CompiletimeOpsStringModuleClass) name match {
               case tpnme.Plus if nArgs == 2 => constantFold2(stringValue, _ + _)
+              case tpnme.Length if nArgs == 1 => constantFold1(stringValue, _.length)
+              case tpnme.Matches if nArgs == 2 => constantFold2(stringValue, _ matches _)
+              case tpnme.Substring if nArgs == 3 =>
+                constantFold3(stringValue, intValue, intValue, (s, b, e) => s.substring(b, e))
```

Comment:
Could we encounter a `IndexOutOfBoundsException`s here? It's probably better to throw a `TypeError` in that case.

Reply:
Yes, I'll change it so that all `constantFoldXX` will trap exceptions from executing their evaluation function and change it to a type error.